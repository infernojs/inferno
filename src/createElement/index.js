import { createVNode } from '../core/createBlueprint';
import { isAttrAnEvent, isArray, isNullOrUndefined, isFunction, isInvalidNode, isAttrAComponentHook, isAttrAHook } from './../core/utils';

export function createAttrsAndEvents(props, tag) {
	let events = null;
	let hooks = null;
	let attrs = null;
	let className = null;
	let style = null;

	if (!isNullOrUndefined(props)) {
		if (isArray(props)) {
			return props;
		}
		for (let prop in props) {
			if (prop === 'className') {
				className = props[prop];
			} else if (prop === 'style') {
				style = props[prop];
			} else if (isAttrAHook(prop) && !isFunction(tag)) {
				if (isNullOrUndefined(hooks)) {
					hooks = {};
				}
				hooks[prop.substring(2).toLowerCase()] = props[prop];
				delete props[prop];
			} else if (isAttrAnEvent(prop) && !isFunction(tag)) {
				if (isNullOrUndefined(events)) {
					events = {};
				}
				events[prop.toLowerCase()] = props[prop];
				delete props[prop];
			} else if (isAttrAComponentHook(prop) && isFunction(tag)) {
				if (isNullOrUndefined(hooks)) {
					hooks = {};
				}
				hooks['c' + prop.substring(3)] = props[prop];
				delete props[prop];
			} else if (!isFunction(tag)) {
				if (isNullOrUndefined(attrs)) {
					attrs = {};
				}
				attrs[prop] = props[prop];
			} else {
				attrs = props;
			}
		}
	}
	return { attrs, events, className, style, hooks };
}

function createChild({ tag, attrs, children, className, style, events, hooks }) {
	if (tag === undefined && !isNullOrUndefined(attrs) && !attrs.tpl && !isNullOrUndefined(children) && children.length === 0) {
		return null;
	}
	const key = !isNullOrUndefined(attrs) && !isNullOrUndefined(attrs.key) ? attrs.key : undefined;

	if (!isNullOrUndefined(children) && children.length === 0) {
		children = null;
	} else if (!isInvalidNode(children)) {
		children = isArray(children) && children.length === 1 ? createChildren(children[0]) : createChildren(children);
	}

	if (key !== undefined) {
		delete attrs.key;
	}
	const attrsAndEvents = createAttrsAndEvents(attrs, tag);
	const vNode = createVNode();

	className = className || attrsAndEvents.className;
	style = style || attrsAndEvents.style;

	vNode.tag = tag || null;
	vNode.attrs = attrsAndEvents.attrs || null;
	vNode.events = attrsAndEvents.events || null;
	vNode.hooks = attrsAndEvents.hooks || null;
	vNode.children = children === undefined ? null : children;
	vNode.key = key === undefined ? null : key;
	vNode.className = className === undefined ? null : className;
	vNode.style = style === undefined ? null : style;

	return vNode;
}

export function createChildren(children) {
	const childrenDefined = !isNullOrUndefined(children);
	if (childrenDefined && isArray(children)) {
		const newChildren = [];

		for (let i = 0; i < children.length; i++) {
			const child = children[i];
			if (!isNullOrUndefined(child) && typeof child === 'object') {
				if (isArray(child)) {
					if (child.length > 0) {
						newChildren.push(createChildren(child));
					} else {
						newChildren.push(null);
					}
				} else {
					newChildren.push(createChild(child));
				}
			} else {
				newChildren.push(child);
			}
		}
		return newChildren;
	} else if (childrenDefined && typeof children === 'object') {
		return children.dom === undefined ? createChild(children) : children;
	}
	return children;
}

export default function createElement(tag, props, ...children) {
	return createChild({ tag, attrs: props, children });
}
