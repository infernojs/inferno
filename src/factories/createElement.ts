import {
	Props,
	InfernoChildren,
	VNodeFlags,
	VNode
} from '../core/shapes';
import { createVNode } from 'inferno';
import {
	isAttrAnEvent,
	isString,
	isInvalid,
	isUndefined,
	isObject,
	isStatefulComponent
} from '../shared';

const componentHooks = {
	onComponentWillMount: true,
	onComponentDidMount: true,
	onComponentWillUnmount: true,
	onComponentShouldUpdate: true,
	onComponentWillUpdate: true,
	onComponentDidUpdate: true
};

export default function createElement<T>(
	name: string | Function,
	props?: T & Props,
	..._children: InfernoChildren[]
): VNode {
	if (isInvalid(name) || isObject(name)) {
		throw new Error('Inferno Error: createElement() name paramater cannot be undefined, null, false or true, It must be a string, class or function.');
	}
	let children: any = _children;
	let ref = null;
	let key = null;
	let events = null;
	let flags = 0;

	if (_children) {
		if (_children.length === 1) {
			children = _children[0];
		} else if (_children.length === 0) {
			children = undefined;
		}
	}
	if (isString(name)) {
		flags = VNodeFlags.HtmlElement;

		switch (name) {
			case 'svg':
				flags = VNodeFlags.SvgElement;
				break;
			case 'input':
				flags = VNodeFlags.InputElement;
				break;
			case 'textarea':
				flags = VNodeFlags.TextareaElement;
				break;
			case 'select':
				flags = VNodeFlags.SelectElement;
				break;
			default:
		}
		for (let prop in props) {
			if (prop === 'key') {
				key = props.key;
				delete props.key;
			} else if (prop === 'children' && isUndefined(children)) {
				children = props.children; // always favour children args, default to props
			} else if (prop === 'ref') {
				ref = props.ref;
			} else if (isAttrAnEvent(prop)) {
				if (!events) {
					events = {};
				}
				events[prop] = props[prop];
				delete props[prop];
			}
		}
	} else {
		flags = isStatefulComponent(name) ? VNodeFlags.ComponentClass : VNodeFlags.ComponentFunction;
		if (!isUndefined(children)) {
			if (!props) {
				props = {} as T;
			}
			props.children = children;
			children = null;
		}
		for (let prop in props) {
			if (componentHooks[prop as string]) {
				if (!ref) {
					ref = {};
				}
				ref[prop] = props[prop];
			} else if (prop === 'key') {
				key = props.key;
				delete props.key;
			}
		}
	}
	return createVNode(
		flags,
		name,
		props,
		children,
		events,
		key,
		ref
	);
}
