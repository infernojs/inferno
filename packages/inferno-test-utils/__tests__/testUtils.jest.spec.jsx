import createElement from "inferno-create-element";
import TestUtils from "inferno-test-utils";

const FunctionalComponent = function(props) {
  return createElement("div", props);
};

describe("renderToSnapshot", () => {
  it("should return a snapshot from a valid vNode", () => {
    const snapshot = TestUtils.renderToSnapshot(
      <FunctionalComponent foo="bar" />
    );

    if (usingJest) {
      expect(snapshot).toMatchSnapshot();
    } else {
      expect(snapshot.props.foo).toBeDefined();
      expect(snapshot.props.foo).toBe("bar");
    }
  });

  it("should return a snapshot with className prop", () => {
    const TestComponent = () => <div className="test">Test</div>;

    const snapshot = TestUtils.renderToSnapshot(<TestComponent />);

    if (usingJest) {
      expect(snapshot).toMatchSnapshot();
    } else {
      expect(snapshot.props.className).toBe("test");
    }
  });
});
