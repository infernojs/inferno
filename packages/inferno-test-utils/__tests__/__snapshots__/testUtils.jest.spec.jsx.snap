// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderToSnapshot should return a snapshot from a valid vNode 1`] = `
<div
  foo="bar"
/>
`;

exports[`renderToSnapshot should return a snapshot with className prop 1`] = `
<div
  className="test"
>
  Test
</div>
`;
